src/audio_pixel.cpp

```diff
@@ -0,0 +1,183 @@
+}
+
+audio_pixel_block_t audio_pixel_mipmap_t::interpolate_block(double t0, double t1, double new_pps, double nearest_pps) {
+    // interpolate from the block with nearest-pps to create a new block at new_pps
+    // the tricky part here is keeping track of the which samples from the nearest block are towards us
+    audio_pixel_block_t nearest_block = m_block_pps[nearest_pps];
+    std::vector<std::vector<audio_pixel_t>> nearest_channel_pixels = nearest_block.get_pixels(t0, t1);
+    int new_number_pixels = (int)round((t1 - t0) * new_pps);
+
+    double new_t_unit = 1/new_pps, nearest_t_unit = 1/nearest_pps; // new/nearest block's time unit (the time between pixels)
+    double nearest_t0 = 0, nearest_t1 = nearest_t_unit; // nearest block's time range and index (position in time of nearest samples)
+    audio_pixel_t nearest_pixel0, nearest_pixel1;
+    int nearest_pixel_idx0 = 0, nearest_pixel_idx1 = 1; 
+    double curr_t = 0; // new block's time 
+
+    std::vector<std::vector<audio_pixel_t>> new_block;
+
+    for (auto& nearest_pixels : nearest_channel_pixels) {
+        // nearest pixels of the specfic channel 
+        audio_pixel_t curr_audio_pixel;
+        std::vector<audio_pixel_t> curr_channel_pixels;
+
+        for (int i = 0; i < new_number_pixels; i++) {
+            curr_t = i * new_t_unit;
+            curr_audio_pixel = audio_pixel_t();
+            
+            // grab the nearest audio pixels and their times
+            nearest_pixel_idx0 = floor(nearest_pps * curr_t);
+            nearest_pixel_idx1 = ceil(nearest_pps * curr_t);
+            nearest_t0 = nearest_pixel_idx0 * nearest_t_unit;
+            nearest_t1 = nearest_t0 + nearest_t_unit;
+            nearest_pixel0 = nearest_pixels[nearest_pixel_idx0];
+            nearest_pixel1 = nearest_pixels[nearest_pixel_idx0];
+
+            // perform liner interpolation for each field of the audio pixel
+            curr_audio_pixel.m_max = linear_interp(curr_t, nearest_t0, nearest_t1, nearest_pixel0.m_max, nearest_pixel0.m_max);
+            curr_audio_pixel.m_min = linear_interp(curr_t, nearest_t0, nearest_t1, nearest_pixel0.m_min, nearest_pixel0.m_min);
+            curr_audio_pixel.m_rms = linear_interp(curr_t, nearest_t0, nearest_t1, nearest_pixel0.m_rms, nearest_pixel0.m_rms);
+
+            curr_channel_pixels.push_back(curr_audio_pixel);
+        }
+        new_block.push_back(curr_channel_pixels);
+    }
+    return audio_pixel_block_t(new_pps, new_block);
+}
+
+double audio_pixel_mipmap_t::closet_val(double val1, double val2, double target) {
+    // val 1 must be greater than val 2
+    if (target - val1 >= val2 - target)
+        return val2;
+    else
+        return val1;
+}
+
+ double audio_pixel_mipmap_t::get_nearest_block_pps(double pix_per_s) {
+    // returns the nearest blocks for a given pixels per seconds using a binary search
+    // assumes that we can get m_block_ppps in sorted form after construction of the map
+    if (m_block_pps.size() == 0)
```

Comment:
I'm starting to think that we should go from `pps` (pixels per second) to `res` (resolution), where `res` is an int corresponding to the number of audio samples per pixel. It makes the code much easier to think about when reading it. I can be the one to do this change.

Reply:
i agree I think that's a great idea